internal/handlers/common/distinct_iterator.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"errors"
+	"sync/atomic"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// DistinctIterator returns an iterator that returns a single document containing
+// the distinct values from documents on the specified key.
+// It will be added to the given closer.
+//
+// Next method return the distinct value of all documents in iterator.
+//
+// Close method closes the underlying iterator.
+func DistinctIterator(iter types.DocumentsIterator, closer *iterator.MultiCloser, key string) (types.DocumentsIterator, error) {
+	res := &distinctIterator{
+		iter: iter,
+	}
```

Comment:
I think we can check path error, then initialize `distinctIterator`, so this can be moved to line L43, right before closer. I find having closer and iterator together is easier to read